runtime/rococo/src/propose_parachain.rs

```diff
@@ -0,0 +1,383 @@
+		///
+		/// This requires:
+		/// - `para_id`: The id of the parachain.
+		/// - `name`: The name of the parachain.
+		/// - `validation_function`: The wasm runtime of the parachain.
+		/// - `initial_head_state`: The genesis state of the parachain.
+		/// - `validators`: Validators that will validate for the relay chain, needs to be at least one.
+		/// - `balance`: The initial balance of the parachain on the relay chain.
+		///
+		/// It will reserve a deposit from the sender account over the lifetime of the chain.
+		#[weight = 1_000_000]
+		fn propose_parachain(
+			origin,
+			para_id: ParaId,
+			name: Vec<u8>,
+			validation_code: ValidationCode,
+			genesis_head: HeadData,
+			validators: Vec<T::ValidatorId>,
+			balance: BalanceOf<T>,
+		) {
+			let who = ensure_signed(origin)?;
+
+			ensure!(name.len() <= T::MaxNameLength::get() as usize, Error::<T>::NameTooLong);
+			ensure!(validators.len() > 0, Error::<T>::AtLeastOneValidatorRequired);
+			ensure!(!Proposals::<T>::contains_key(&para_id), Error::<T>::ParachainIdAlreadyProposed);
+			ensure!(
+				!runtime_parachains::paras::Module::<T>::parachains().contains(&para_id),
+				Error::<T>::ParachainIdAlreadyTaken,
+			);
+			ensure!(
+				!runtime_parachains::paras::Module::<T>::upcoming_paras().contains(&para_id),
+				Error::<T>::ParachainIdAlreadyTaken,
+			);
+			ensure!(validation_code.0.starts_with(runtime_common::WASM_MAGIC), Error::<T>::DefinitelyNotWasm);
+
+			let active_validators = Session::<T>::validators();
+			ensure!(
+				validators.iter().all(|v| !active_validators.contains(v)),
+				Error::<T>::ValidatorAlreadyRegistered,
+			);
+			Proposals::<T>::iter().try_for_each(|(_, prop)|
+				if validators.iter().all(|v| !prop.validators.contains(v)) {
+					Ok(())
+				} else {
+					Err(Error::<T>::ValidatorAlreadyRegistered)
+				}
+			)?;
+
+			pallet_balances::Module::<T>::reserve(&who, T::ProposeDeposit::get())?;
+
+			let proposal = Proposal {
+				name: name.clone(),
+				proposer: who,
+				validators: validators.into(),
+				genesis_head,
+				validation_code,
+				balance,
+			};
+
+			Proposals::<T>::insert(para_id, proposal);
+
+			Self::deposit_event(Event::ParachainProposed(name, para_id));
+		}
+
+		/// Approve a parachain proposal.
+		#[weight = 100_000]
+		fn approve_proposal(
+			origin,
+			para_id: ParaId,
+		) {
+			T::PriviledgedOrigin::ensure_origin(origin)?;
+
+			ensure!(Proposals::<T>::contains_key(&para_id), Error::<T>::ProposalNotFound);
+
+			Self::is_approved_or_scheduled(para_id)?;
+
+			ApprovedProposals::append(para_id);
+
+			Self::deposit_event(Event::ParachainApproved(para_id));
+		}
+
+		/// Cancel a parachain proposal.
+		///
+		/// This also unreserves the deposit.
+		#[weight = 100_000]
+		fn cancel_proposal(origin, para_id: ParaId) {
+			let who = match EnsurePriviledgedOrSigned::<T>::ensure_origin(origin)? {
+				Either::Left(_) => None,
+				Either::Right(who) => Some(who),
+			};
+
+			Self::is_approved_or_scheduled(para_id)?;
```

Comment:
This seems like it should be the last check among these checks since there is an iterator inside